include/ocpp/common/message_queue.hpp

```diff
@@ -342,44 +345,52 @@ template <typename M> class MessageQueue {
                 // prioritize the message with the oldest timestamp
                 auto now = DateTime();
                 std::shared_ptr<ControlMessage<M>> message = nullptr;
+                auto selected_normal_message_it = normal_message_queue.end(); // Use as a placeholder
                 QueueType queue_type = QueueType::None;
 
-                if (!this->normal_message_queue.empty()) {
-                    auto& normal_message = this->normal_message_queue.front();
-                    EVLOG_debug << "normal msg timestamp: " << normal_message->timestamp;
-                    if (normal_message->timestamp <= now) {
-                        EVLOG_debug << "normal message timestamp <= now";
-                        message = normal_message;
-                        queue_type = QueueType::Normal;
-                    } else {
-                        EVLOG_error << "A normal message should not have a timestamp in the future: "
-                                    << normal_message->timestamp << " now: " << now;
+                for (auto normal_message_it = normal_message_queue.begin();
```

Comment:
Seems like this searching for a message is kind of duplicated, once for normal and once for transactions. Could be a good candidate to move into a helper function I would say.

Also, first checking for transaction message and then checking the other queue would make sense. Now we always check the normal queue and then possibly overwrite using the transaction queue.